src/ImmutableProperties.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Codelicia\Immutable;
+
+use ReflectionObject;
+use ReflectionProperty;
+use ReflectionType;
+use RuntimeException;
+use function array_key_exists;
+use function class_implements;
+use function get_class;
+use function gettype;
+use function in_array;
+use function interface_exists;
+use function is_object;
+use function spl_object_hash;
+use function sprintf;
+
+trait ImmutableProperties
+{
+    /** @return string[][]|ReflectionProperty[][] */
+    protected function &reflectionProperty(): array
+    {
+        static $reflectionProperty = [];
+
+        if (! array_key_exists(spl_object_hash($this), $reflectionProperty)) {
+            $reflectionProperty[spl_object_hash($this)] = [];
+        }
+
+        return $reflectionProperty[spl_object_hash($this)];
+    }
+
+    public function __construct()
+    {
+        $this->init();
+    }
+
+    public function init(): void
+    {
+        $reflection = new ReflectionObject($this);
+        $store = &$this->reflectionProperty();
+        $defaultProperties = $reflection->getDefaultProperties();
+
+        foreach ($reflection->getProperties(ReflectionProperty::IS_PUBLIC) as $reflectionProperty) {
+            $propertyName = $reflectionProperty->name;
+            /** @var \ReflectionType $type */
+            $type = $reflectionProperty->getType();
+
+            $store[$propertyName] = [
+                'reflection' => $reflectionProperty,
+                'value' => $defaultProperties[$propertyName] ?: null,
+                'isInitialized' => isset($defaultProperties[$propertyName]) ? true : $reflectionProperty->isInitialized($this),
+            ];
+
+            if (isset($defaultProperties[$propertyName]) || $type->allowsNull()) {
+                unset($this->{$propertyName});
+            }
+        }
+    }
+
+    public function __set(string $prop, $value)
+    {
+        $ref = &$this->reflectionProperty();
+        $property = $ref[$prop];
+
+        if ($property['isInitialized']) {
+            // @TODO throws specific exception?
+            throw new RuntimeException(sprintf('Cannot reassign value to property "%s"', $prop));
+        }
+
+        /** @var ReflectionProperty $reflection */
+        $reflection = $property['reflection'];
+
+        if ($reflection->hasType()) {
```

Comment:
I think this entire block can be skipped: the property would reject the assignment anyway, no?